### PR TITLE
SJRK-205: Update "Delete story" server call to remove associated files as well

### PR DESCRIPTION
**Pull request by BlueSlug**
As SJRK-184 is still open currently, if a story is uploaded with blocks containing duplicate file names, the first one is tracked and subsequent files are not, therefore upon story deletion these unreferenced files will not be deleted.

### CODE REVIEW

File: src/js/requestHandlers.js [hunk 1]
```diff
@@ -230,6 +234,32 @@ sjrk.storyTelling.server.deleteStoryFromCouch = function (id, deleteStoryDataSou
...
+
+    // remove duplicate entries so we don't try to delete already-deleted files
+    filesToDelete = filesToDelete.filter(function (fileName, index, self) {
+        return self.indexOf(fileName) === index;
+    });
+
+    fluid.each(filesToDelete, function (fileToDelete) {
+        var filePath = "./" + uploadedFilesHandlerPath + "/" + fileToDelete;
+
+        fs.unlink(filePath, function (err) {
```
Comment (simonbates) [inline src/js/requestHandlers.js]: The way that async `fs.unlink` is used here will detach the deletion of the files from the response. We should either use a promise to stay connected to the response, or switch to the `fs.unlinkSync` version.

File: src/js/requestHandlers.js [hunk 2]
```diff
@@ -230,6 +234,32 @@ sjrk.storyTelling.server.deleteStoryFromCouch = function (id, deleteStoryDataSou
...
+        }
+    });
+
+    // remove duplicate entries so we don't try to delete already-deleted files
+    filesToDelete = filesToDelete.filter(function (fileName, index, self) {
+        return self.indexOf(fileName) === index;
+    });
+
+    fluid.each(filesToDelete, function (fileToDelete) {
+        var filePath = "./" + uploadedFilesHandlerPath + "/" + fileToDelete;
```
Comment (simonbates) [inline src/js/requestHandlers.js]: I believe that we have a security vulnerability here: if I can submit a story with a media URL containing "../", I can delete any file that the node process has access to. Proof of concept:

```
PS C:\Users\simon\Scratch\a> dir


    Directory: C:\Users\simon\Scratch\a


Mode                LastWriteTime         Length Name
----                -------------         ------ ----
d-----       2019-01-25   9:16 AM                b
-a----       2019-01-25   9:17 AM             16 hello.txt


PS C:\Users\simon\Scratch\a> node
> fs.unlinkSync("./b/../hello.txt")
undefined
>
(To exit, press ^C again or type .exit)
>
PS C:\Users\simon\Scratch\a> dir


    Directory: C:\Users\simon\Scratch\a


Mode                LastWriteTime         Length Name
----                -------------         ------ ----
d-----       2019-01-25   9:16 AM                b


PS C:\Users\simon\Scratch\a>
```

File: src/js/requestHandlers.js (hunk 1 above)
Comment (simonbates) [reply src/js/requestHandlers.js]: And as I suggested yesterday, I would recommend moving the files to a 'trash' directory, such as `deleted_uploads`, rather than deleting. So that we have some recovery potential.

File: src/js/requestHandlers.js (hunk 2 above)
Comment (simonbates) [reply src/js/requestHandlers.js]: Thinking about options here.

We could use [`path.basename`](https://nodejs.org/api/path.html#path_path_basename_path_ext) prior to constructing the path to operate on.

But I'm thinking that instead we do a test (say using a regular expression) that the media URL contains only exactly the characters that we expect -- we construct the names used here, so we can reliably test. I think this would be:

- upper and lower case letters (unless we normalize the extension to lower case?)
- numbers
- "-"
- "."

Comment (simonbates) [reply src/js/requestHandlers.js]: Let's add a new function `isValidMediaFilename` that checks filename contents as above and can also test for other cases such as `null`, `undefined`, empty string, and such. And if it's a function, we can unit test it.

Comment (BlueSlug) [reply src/js/requestHandlers.js]: Excellent catch! I've added checks to ensure:
- the filename is truthy (covers `null`, `undefined`, empty string, etc.)
- the filename is a valid UUID (all stored files should be)
- the file can be accessed for read and write in the uploads directory

File: src/js/requestHandlers.js (hunk 1 above)
Comment (BlueSlug) [reply src/js/requestHandlers.js]: The promises functions in the FileSystem API are still experimental, so I will avoid using those for the time being and use the synchronous functions instead. I've added code that moves the given files to a new `deleted_uploads` directory as you suggested, then ensures the file has been moved, is still accessible by using `fs.accessSync` and is no longer accessible at the old path.

File: src/js/requestHandlers.js [hunk 3]
```diff
@@ -230,6 +235,85 @@ sjrk.storyTelling.server.deleteStoryFromCouch = function (id, deleteStoryDataSou
...
+    fluid.each(storyContent, function (block) {
+        var blockFileName = "";
+
+        if (block.blockType === "image") {
+            blockFileName = block.imageUrl;
+        } else if (block.blockType === "audio" || block.blockType === "video") {
+            blockFileName = block.mediaUrl;
+        }
+
+        if (sjrk.storyTelling.server.isValidMediaFilename(blockFileName)) {
```
Comment (simonbates) [inline src/js/requestHandlers.js:278]: You're not passing `serverConfig` to `isValidMediaFilename`. Though, I think it's actually better to remove the `access` check from that function anyway. We have a race condition: there is no guarantee that the file will still exist when we try to move it, so we need to handle the case that the file is missing there anyway.

File: src/js/requestHandlers.js [hunk 4]
```diff
@@ -230,6 +235,85 @@ sjrk.storyTelling.server.deleteStoryFromCouch = function (id, deleteStoryDataSou
...
+    fluid.each(filesToDelete, function (fileToDelete) {
+        sjrk.storyTelling.server.deleteSingleFileRecoverable(fileToDelete, serverConfig);
+    });
+};
+
+sjrk.storyTelling.server.isValidMediaFilename = function (fileName, serverConfig) {
+    try {
+        if (!fileName || fileName === "null" || fileName === "undefined")
+        {
+            throw "The filename is not valid: not truthy.";
```
Comment (simonbates) [inline src/js/requestHandlers.js]: This structure is overly complex: to throw an exception only to catch it within the same function and return `false` -- let's instead just return `false` directly from these tests.

File: src/js/requestHandlers.js [hunk 5]
```diff
@@ -230,6 +235,85 @@ sjrk.storyTelling.server.deleteStoryFromCouch = function (id, deleteStoryDataSou
...
+        }
+
+        if (sjrk.storyTelling.server.isValidUuidFileName(fileName)) {
+            throw "The filename is not valid: does not match UUID file name pattern.";
+        }
+
+        var fileWithFullPath = sjrk.storyTelling.server.getServerPathForFile(fileName, serverConfig.uploadedFilesHandlerPath);
+        fs.accessSync(fileWithFullPath, fs.constants.W_OK | fs.constants.R_OK);
```
Comment (simonbates) [inline src/js/requestHandlers.js]: As discussed above, I would remove this test.

File: src/js/requestHandlers.js [hunk 6]
```diff
@@ -230,6 +235,85 @@ sjrk.storyTelling.server.deleteStoryFromCouch = function (id, deleteStoryDataSou
...
+
+        return true; // it is a valid file name
+    } catch (err) {
+        return false; // it is not a valid file name
+    }
+};
+
+sjrk.storyTelling.server.isValidUuidFileName = function (fileName) {
```
Comment (simonbates) [inline src/js/requestHandlers.js]: The name does not match the function -- it's the inverse.

File: src/js/requestHandlers.js [hunk 7]
```diff
@@ -230,6 +235,85 @@ sjrk.storyTelling.server.deleteStoryFromCouch = function (id, deleteStoryDataSou
...
+    var validUuid = /^\b[0-9a-f]{8}\b-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-\b[0-9a-f]{12}\b/;
+    return !validUuid.test(path.basename(fileName));
```
Comment (simonbates) [inline src/js/requestHandlers.js]: I don't think we want `path.basename` here.

File: src/js/requestHandlers.js [hunk 8]
```diff
@@ -230,6 +235,85 @@ sjrk.storyTelling.server.deleteStoryFromCouch = function (id, deleteStoryDataSou
...
+        fs.accessSync(fileWithFullPath, fs.constants.W_OK | fs.constants.R_OK);
+
+        return true; // it is a valid file name
+    } catch (err) {
+        return false; // it is not a valid file name
+    }
+};
+
+sjrk.storyTelling.server.isValidUuidFileName = function (fileName) {
+    var validUuid = /^\b[0-9a-f]{8}\b-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-\b[0-9a-f]{12}\b/;
```
Comment (simonbates) [inline src/js/requestHandlers.js]: This expression is pretty complex. Could you add some unit tests for this code?

File: src/js/requestHandlers.js [hunk 9]
```diff
@@ -85,12 +87,12 @@ sjrk.storyTelling.server.handleGetStory = function (request, dataSource, uploade
         fluid.transform(response.content, function (block) {
             if (block.blockType === "image") {
                 if (block.imageUrl) {
-                    block.imageUrl = uploadedFilesHandlerPath + "/" + block.imageUrl;
+                    block.imageUrl = uploadedFilesHandlerPath + path.sep + block.imageUrl;
```
Comment (simonbates) [inline src/js/requestHandlers.js]: I believe that `path.sep` is incorrect in this function -- we're building URLs in this function, right? And the separator is always '/'.

File: src/js/requestHandlers.js [hunk 10]
```diff
@@ -230,6 +235,85 @@ sjrk.storyTelling.server.deleteStoryFromCouch = function (id, deleteStoryDataSou
...
+
+sjrk.storyTelling.server.getServerPathForFile = function (fileName, directoryName) {
+    return "." + directoryName + path.sep + fileName;
+};
+
+sjrk.storyTelling.server.deleteSingleFileRecoverable = function (fileToDelete, serverConfig) {
+    var recoveryPath = sjrk.storyTelling.server.getServerPathForFile(fileToDelete, serverConfig.deletedFilesRecoveryPath);
+    var deletionPath = sjrk.storyTelling.server.getServerPathForFile(fileToDelete, serverConfig.uploadedFilesHandlerPath);
+
+    fs.renameSync(deletionPath, recoveryPath); // move it to the recovery dir
```
Comment (simonbates) [inline src/js/requestHandlers.js]: Add error handling here.

File: src/js/requestHandlers.js (hunk 1 above)
Comment (simonbates) [reply src/js/requestHandlers.js]: We could use Infusion promises. But I think using synchronous file access here is fine. It's simpler and as we don't intend to use this code very much, the performance impact isn't super important.

Comment (simonbates) [reply src/js/requestHandlers.js]: Though it's more accurate to say: we _hope_ to not use this code very much.

File: src/js/requestHandlers.js (hunk 6 above)
Comment (simonbates) [reply src/js/requestHandlers.js]: Though, I think once `isValidMediaFilename` is simplified, we can fold this function's behaviour directly into that function. And remove this separate function.

File: src/js/requestHandlers.js (hunk 4 above)
Comment (simonbates) [reply src/js/requestHandlers.js]: Are these tests against "null" and "undefined" necessary? Won't they fail the regular expression test anyway?

File: src/js/requestHandlers.js (hunk 7 above)
Comment (BlueSlug) [reply src/js/requestHandlers.js]: I misread the docs for path.basename() and didn't pass in the "ext" parameter which will remove the file extension from the rest of the filename. Would you suggest not using that function anyway?

File: src/js/requestHandlers.js (hunk 9 above)
Comment (BlueSlug) [reply src/js/requestHandlers.js]: Thanks for catching this, I think this one tripped me up in 2017 as well. I got a bit overzealous with my replacement.

**Comment by simonbates** [comment]
I think it's also worth adding some new logic to the file upload handler. At the moment we have a vector by which someone can submit stories with arbitrary URLs -- by submitting media blocks with no matching file uploaded. I would add code to this test in the upload handler:

https://github.com/fluid-project/sjrk-story-telling-server/blob/37fb2bbf66d1098ba9b88e42d49523ab32d8d83b/src/js/requestHandlers.js#L144-L153

Currently, if no `mediaFile` is found, we leave the block's URL as is. We should instead set it to `null`.

File: src/js/requestHandlers.js (hunk 7 above)
Comment (simonbates) [reply src/js/requestHandlers.js]: Yes, I think we should remove the use of `path.basename` and verify that the filename matches exactly the format we want (and that it doesn't have any path parts).

File: src/js/requestHandlers.js [hunk 11]
```diff
@@ -230,6 +235,83 @@ sjrk.storyTelling.server.deleteStoryFromCouch = function (id, deleteStoryDataSou
...
+/*
+ * Verifies that a given file name follows the UUID format as laid out in
+ * RFC4122. A detailed description of the format can be found here:
+ * https://en.wikipedia.org/wiki/Universally_unique_identifier#Format
+ */
+sjrk.storyTelling.server.isValidMediaFilename = function (fileName) {
+    if (fileName && typeof fileName === "string" ) {
+        var fileNameWithoutExtension = path.basename(fileName, path.extname(fileName));
+
+        var validUuid = /\b[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}\b/;
```
Comment (simonbates) [inline src/js/requestHandlers.js]: Lets test the whole name with `^` and `$` and remove the use of `\b` -- let's test the whole filename, rather than testing a substring.

File: src/js/requestHandlers.js [hunk 12]
```diff
@@ -230,6 +235,83 @@ sjrk.storyTelling.server.deleteStoryFromCouch = function (id, deleteStoryDataSou
...
+};
+
+/*
+ * Verifies that a given file name follows the UUID format as laid out in
+ * RFC4122. A detailed description of the format can be found here:
+ * https://en.wikipedia.org/wiki/Universally_unique_identifier#Format
+ */
+sjrk.storyTelling.server.isValidMediaFilename = function (fileName) {
+    if (fileName && typeof fileName === "string" ) {
+        var fileNameWithoutExtension = path.basename(fileName, path.extname(fileName));
```
Comment (simonbates) [inline src/js/requestHandlers.js]: Please remove use of `path.basename`.

File: tests/testServerWithStorage.js [hunk 13]
```diff
@@ -487,6 +487,52 @@ sjrk.storyTelling.server.testServerWithStorageDefs.testImageRetrieval = function
     jqUnit.assertEquals("header.content-length is 3719", "3719", request.nativeResponse.headers["content-length"]);
 };
 
+jqUnit.test("Test isValidMediaFilename function", function () {
+    jqUnit.expect(15);
+
+    var inputFileNames = [
```
Comment (simonbates) [inline tests/testServerWithStorage.js]: I think this would be clearer if we combined the test cases and expected value in a single array.

File: tests/testServerWithStorage.js [hunk 14]
```diff
@@ -487,6 +487,52 @@ sjrk.storyTelling.server.testServerWithStorageDefs.testImageRetrieval = function
...
+        [],
+        [0],
+        "",
+        "FailingFileName",
+        "FailingFileName.ext",
+        "1f4EAE4020CF11E9975C2103755D20B8.mp4",
+        "1f4eae4020cf11e9975c2103755d20b8.mp4",
+        "jpg.1f4845a0-20cf-11e9-975c-2103755d20b8",
+        "1f4845a0-20cf-11e9-975c-2103755d20b8.jpg",
+        "/uploads/10c4e170-20d2-11e9-ad76-451a41995405.jpg",
```
Comment (simonbates) [inline tests/testServerWithStorage.js]: This should be a failure, rather than a pass. And should be, once we remove the call to `path.basename`.

File: src/js/requestHandlers.js [hunk 15]
```diff
@@ -230,6 +235,83 @@ sjrk.storyTelling.server.deleteStoryFromCouch = function (id, deleteStoryDataSou
...
+    return "." + directoryName + path.sep + fileName;
+};
+
+sjrk.storyTelling.server.deleteSingleFileRecoverable = function (fileToDelete, serverConfig) {
+    var recoveryPath = sjrk.storyTelling.server.getServerPathForFile(fileToDelete, serverConfig.deletedFilesRecoveryPath);
+    var deletionPath = sjrk.storyTelling.server.getServerPathForFile(fileToDelete, serverConfig.uploadedFilesHandlerPath);
+
+    // make sure the file is in the uploads dir before we begin
+    try {
+        fs.accessSync(deletionPath, fs.constants.W_OK | fs.constants.R_OK);
```
Comment (simonbates) [inline src/js/requestHandlers.js]: We don't need this test. Just let the rename fail.

File: tests/testServerWithStorage.js [hunk 16]
```diff
@@ -487,6 +487,52 @@ sjrk.storyTelling.server.testServerWithStorageDefs.testImageRetrieval = function
...
+        "1f4845a0-20cf-11e9-975c-2103755d20b8.somethingVeryLong",
+    ];
```
Comment (simonbates) [inline tests/testServerWithStorage.js:515]: Let's add a failing testcase with a `"../"`, like: `"../1f4845a0-20cf-11e9-975c-2103755d20b8.jpg"`.

File: src/js/serverSetup.js [hunk 17]
```diff
@@ -22,8 +22,8 @@ fluid.defaults("sjrk.storyTelling.server", {
                 globalConfig: {
                     binaryUploadDirectory: "./uploads",
                     uploadedFilesHandlerPath: "/uploads",
+                    deletedFilesRecoveryPath: "/deleted_uploads",
```
Comment (cindyli) [inline src/js/serverSetup.js:25]: Is there a JIRA created for adding tests for this new endpoint?

File: src/js/requestHandlers.js [hunk 18]
```diff
@@ -181,14 +192,14 @@ fluid.defaults("sjrk.storyTelling.server.deleteStoryHandler", {
     invokers: {
         handleRequest: {
             funcName: "sjrk.storyTelling.server.handleDeleteStory",
-            args: ["{arguments}.0", "{server}.deleteStoryDataSource", "{server}.storyDataSource"]
+            args: ["{arguments}.0", "{server}.deleteStoryDataSource", "{server}.storyDataSource", "{server}.options.globalConfig"]
```
Comment (cindyli) [inline src/js/requestHandlers.js]: `{server}.options.globalConfig` is given to the top function and being passed down deeply to reach sjrk.storyTelling.server.deleteSingleFileRecoverable() where it eventually gets used.

The same issue happens for other 2 arguments `{server}.deleteStoryDataSource` and `{server}.storyDataSource` where they are only needed by another function called by the top function. 

To clarify the logic, this top function seems only need one argument "{that}.req.params.id". Other functions that require extra arguments can be defined as invokers and have proper arguments supplied via their own args lists.

File: src/js/serverSetup.js (hunk 17 above)
Comment (BlueSlug) [reply src/js/serverSetup.js:25]: I have filed a Jira to address this: https://issues.fluidproject.org/browse/SJRK-208

Please feel free to make any suggestions if it does not cover what you imagined!

File: src/js/requestHandlers.js (hunk 18 above)
Comment (BlueSlug) [reply src/js/requestHandlers.js]: I believe I've achieved what you wanted. I'm still a little unsure about passing in the `request` object into `sjrk.storyTelling.server.handleDeleteStory` instead of passing the story ID and the success and error events separately, so I've left that part as it was. Is that more or less what you were thinking?

File: src/js/requestHandlers.js [hunk 19]
```diff
@@ -192,14 +192,25 @@ fluid.defaults("sjrk.storyTelling.server.deleteStoryHandler", {
     invokers: {
         handleRequest: {
             funcName: "sjrk.storyTelling.server.handleDeleteStory",
-            args: ["{arguments}.0", "{server}.deleteStoryDataSource", "{server}.storyDataSource", "{server}.options.globalConfig"]
+            args: ["{that}", "{arguments}.0"]
```
Comment (cindyli) [inline src/js/requestHandlers.js]: "{arguments}.0" should point to "{that}" as well. Passing only one of them should be enough.

File: src/js/requestHandlers.js [hunk 20]
```diff
@@ -230,6 +254,74 @@ sjrk.storyTelling.server.deleteStoryFromCouch = function (id, deleteStoryDataSou
...
+
+sjrk.storyTelling.server.deleteSingleFileRecoverable = function (fileToDelete, deletedFilesRecoveryPath, uploadedFilesHandlerPath) {
+    var recoveryPath = sjrk.storyTelling.server.getServerPathForFile(fileToDelete, deletedFilesRecoveryPath);
+    var deletionPath = sjrk.storyTelling.server.getServerPathForFile(fileToDelete, uploadedFilesHandlerPath);
+
+    // move it to the recovery dir and make sure it was moved
+    try {
+        fs.renameSync(deletionPath, recoveryPath);
+        fs.accessSync(recoveryPath, fs.constants.W_OK | fs.constants.R_OK);
+        fluid.log("Moved file to recovery dir:", recoveryPath);
```
Comment (cindyli) [inline src/js/requestHandlers.js:322]: It would be helpful to provide the info of the source file location `deletionPath` in the message.

When deleted files are moved to a recovery directory, there are no side information saved along with these files such as where they come from or which story they belonged to, etc. This makes me wonder what's the purpose of having deleted files archived? For future recovery? Or security concerns to trace back? Also, when more and more files are accumulated in the recovery folder, we should think about the archive and clean-up rules for this folder. This doesn't need to be addressed here but it's probably worthwhile to note it as a to-do.

File: src/js/requestHandlers.js [hunk 21]
```diff
@@ -230,6 +254,74 @@ sjrk.storyTelling.server.deleteStoryFromCouch = function (id, deleteStoryDataSou
...
+    } catch (err) {
+        fluid.fail("Error moving file to recover dir:", deletionPath, "Error detail:", err.toString());
```
Comment (cindyli) [inline src/js/requestHandlers.js]: typo: recover -> recovery. It might be more clearer to write this message as:

"Error moving file ", deletionPath, " to recovery dir. Error detail: ", err.toString()

**Comment by simonbates** [comment]
We'll want to update the Docker container filesystem so that the new `deleted_uploads` directory is managed like `uploads`: as a directory mounted from the host filesystem. So that it survives container redeployments.

Configuration for managing the directories:

- https://github.com/waharnum/sjrk-story-telling-server/blob/master/docker-compose.cloud.yml
- https://github.com/inclusive-design/ci-service/blob/master/jenkins_jobs/stack-stories.floeproject.org.yml

File: src/js/requestHandlers.js (hunk 18 above)
Comment (cindyli) [reply src/js/requestHandlers.js]: Sorry, missed this question. Yes, handleDeleteStory() needs the pass in of `request` object for firing success and error events. Only passing the story id is not enough. Your change looks good.

File: src/js/requestHandlers.js [hunk 22]
```diff
@@ -181,14 +192,25 @@ fluid.defaults("sjrk.storyTelling.server.deleteStoryHandler", {
...
-            args: ["{arguments}.0", "{server}.deleteStoryDataSource", "{server}.storyDataSource"]
+            args: ["{that}", "{arguments}.0"]
+        },
+        deleteStoryFromCouch: {
+            funcName: "sjrk.storyTelling.server.deleteStoryFromCouch",
+            args: ["{that}", "{arguments}.0", "{server}.deleteStoryDataSource", "{server}.storyDataSource"]
+        },
+        deleteStoryFiles: {
+            funcName: "sjrk.storyTelling.server.deleteStoryFiles",
+            args: ["{that}", "{arguments}.0"]
```
Comment (simonbates) [inline src/js/requestHandlers.js]: Please add comments for the new invokers, documenting their parameters. Such as:

```
args: ["{that}", "{arguments}.0"] // storyContent
```

File: src/js/requestHandlers.js [hunk 23]
```diff
@@ -230,6 +254,74 @@ sjrk.storyTelling.server.deleteStoryFromCouch = function (id, deleteStoryDataSou
...
+};
+
+/*
+ * Verifies that a given file name follows the UUID format as laid out in
+ * RFC4122. A detailed description of the format can be found here:
+ * https://en.wikipedia.org/wiki/Universally_unique_identifier#Format
+ */
+sjrk.storyTelling.server.isValidMediaFilename = function (fileName) {
+    if (fileName && typeof fileName === "string" ) {
+        var validUuid = /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}\..*$/;
```
Comment (simonbates) [inline src/js/requestHandlers.js]: Could we constrain the extension part a little more: `\.\w+` rather than `\..*`

Are we sure that we will always have an extension?

File: tests/testServerWithStorage.js [hunk 24]
```diff
@@ -487,6 +487,36 @@ sjrk.storyTelling.server.testServerWithStorageDefs.testImageRetrieval = function
...
+        { input: [0], expected: false },
+        { input: "", expected: false },
+        { input: "../", expected: false },
+        { input: "FailingFileName", expected: false },
+        { input: "FailingFileName.ext", expected: false },
+        { input: "1f4EAE4020CF11E9975C2103755D20B8.mp4", expected: false },
+        { input: "1f4eae4020cf11e9975c2103755d20b8.mp4", expected: false },
+        { input: "jpg.1f4845a0-20cf-11e9-975c-2103755d20b8", expected: false },
+        { input: "/uploads/10c4e170-20d2-11e9-ad76-451a41995405.jpg", expected: false },
+        { input: "../10c4e170-20d2-11e9-ad76-451a41995405.jpg", expected: false },
```
Comment (simonbates) [inline tests/testServerWithStorage.js]: For the testcases that fail due to having prepended directories, let's use the same filenames as one of the passing cases. `1f4845a0-20cf-11e9-975c-2103755d20b8.jpg` So that it's clear that the test is failing due to the directory parts, and not because of an issue in the filename.

File: src/js/requestHandlers.js [hunk 25]
```diff
@@ -230,6 +254,74 @@ sjrk.storyTelling.server.deleteStoryFromCouch = function (id, deleteStoryDataSou
...
+
+        if (block.blockType === "image") {
+            blockFileName = block.imageUrl;
+        } else if (block.blockType === "audio" || block.blockType === "video") {
+            blockFileName = block.mediaUrl;
+        }
+
+        if (sjrk.storyTelling.server.isValidMediaFilename(blockFileName)) {
+            filesToDelete.push(blockFileName);
+        }
```
Comment (simonbates) [inline src/js/requestHandlers.js:282]: Probably a good idea to log filenames that we reject.

File: src/js/requestHandlers.js (hunk 23 above)
Comment (BlueSlug) [reply src/js/requestHandlers.js]: Whether we should always expect an extension is an interesting question, as the alternative would mean that people are uploading files with the correct MIME type (`audio/*`, `image/*` or `video/*`) but simultaneously don't have a file extension. I'm having trouble finding the exact implementation used by browsers to see how they determine whether a file matches those MIME types.

It is worth noting, though, that Edge and Firefox for Android don't currently support the `accept` attribute on the file upload controls (https://caniuse.com/#feat=input-file-accept), so it's possible to upload any type of file for any of the block types, including files without extensions. This is certainly a possibility, then. I have made the extension part optional, so a filename will be either only a UUID or a UUID with a file extension.
